libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -553,9 +555,27 @@ proc subscribe*(p: PubSub, topic: string, handler: TopicHandler) {.public.} =
 
   p.updateTopicMetrics(topic)
 
-method publish*(
+method createMessage*(
     p: PubSub, topic: string, data: seq[byte]
-): Future[int] {.base, async: (raises: []), public.} =
+): Result[(Message, MessageId), string] {.base, gcsafe, raises: [].} =
+  let
+    msg =
+      if p.anonymize:
+        Message.init(none(PeerInfo), data, topic, none(uint64), false)
+      else:
+        inc p.msgSeqno
+        Message.init(some(p.peerInfo), data, topic, some(p.msgSeqno), p.sign)
+    msgId = p.msgIdProvider(msg).valueOr:
```

Comment:
As discussed earlier, this is a logic error which should never happen for messages coming from the application, ie the message id function must always be capable of creating an ID for messages that the application wants to publish.

The default message id function never fails, and if the application changes message id function, it should also ensure that it's not creating messages that fail it.

The message id function is allowed to return an error because it's also used for messages coming from untrusted sources, ie it must be robust against malicious and/or invalid data.

As such, this should not be an error return.